sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/MultiCrateTests.java

```diff
@@ -138,14 +142,70 @@ public void testPackagedDemos() throws SQLException, IOException, InterruptedExc
         FilenameFilter filter = (_d, name) -> !name.contains("setup") && name.endsWith(".sql");
         String[] sqlFiles = dir.list(filter);
         assert sqlFiles != null;
+        Arrays.sort(sqlFiles);
+
         for (String sqlFile: sqlFiles) {
             // System.out.println(sqlFile);
             String basename = Utilities.getBaseName(sqlFile);
-            String udf = basename + ".udf.rs";
-            File udfFile = new File(dir.getPath() + "/" + udf);
-            if (!udfFile.exists())
-                this.compileMultiCrate(dir.getPath() + "/" + sqlFile);
+            String originalUdf = basename + ".udf.rs";
+            this.compileToMultiCrate(dir.getPath() + "/" + sqlFile, false);
+
+            File udfFile = new File(dir.getPath() + "/" + originalUdf);
+            File udf = null;
+            if (udfFile.exists()) {
+                String rs = Utilities.readFile(udfFile.getPath());
+                udf = this.createUdfFile(rs);
+
+                File cargo = new File(dir.getPath() + "/" + basename + ".udf.toml");
+                if (cargo.exists()) {
+                    String cargoContents = Utilities.readFile(cargo.getPath());
+                    this.appendCargoDependencies(cargoContents);
+                }
+            }
+            Utilities.compileAndCheckRust(BaseSQLTests.RUST_CRATES_DIRECTORY, true);
+            if (udf != null) {
+                //noinspection ResultOfMethodCallIgnored
+                udf.delete();
+            }
+        }
+    }
+
+    void appendCargoDependencies(String source) throws IOException {
+        Path dir = Paths.get(BaseSQLTests.RUST_CRATES_DIRECTORY, MultiCrates.FILE_PREFIX + "x_globals");
+        File cargo = new File(dir.toFile(), "Cargo.toml");
+        String[] lines = source.split("\n");
+        FileWriter writer = new FileWriter(cargo, true);
+        boolean append = false;
+        for (String line: lines) {
+            if (append) {
+                writer.append(line);
+                writer.append(System.lineSeparator());
+            }
+            if (line.equals("[dependencies]"))
+                append = true;
         }
+        writer.close();
+    }
+
+    @Test
+    public void issue3903() throws IOException, SQLException, InterruptedException {
+        String url = "https://raw.githubusercontent.com/feldera/techdemo-spreadsheet/refs/heads/main/feldera/program.sql";
```

Comment:
still don't think its a good idea to hardcode this URL. ideally we have tests that still work in a year even if this changes (could be just some refactoring that mixes the path, now your test-suite will break for all commits going forward and CI will just fail). IMO better to just include this particular SQL version in our code.

Reply:
on the contrary, I really want to compile the latest version of the spreadsheet.
if the path is changed, we should fix the test.

Reply:
we do the same for the demos, but use the checked out repo

Reply:
the demos is different because it's in the same repo. we shouldn't rely on the structure of a different repo to run the tests of this repo (even if it's our repo)... 

or another point against doing this: clearly this fixes a particular issue about ports .. how do you guarantee that a future revision of the spreadsheet code still exercises the same code-path/bug

Reply:
can we add this as a packaged demo?

Reply:
yes

Reply:
this won't guarantee that bug is exercised...
do I just copy the sql/udf/cargo files?
I can do it in this PR

Reply:
I am also not sure what people will see when they run it

Reply:
>do I just copy the sql/udf/cargo files?
I think so